core/src/main/java/org/apache/druid/segment/column/Types.java

```diff
@@ -112,4 +121,529 @@
+   */
+  @Nullable
+  public static byte[] readNullableVariableBlob(ByteBuffer buffer, int offset)
+  {
+    // | null (byte) | length (int) | bytes |
+    final int length = buffer.getInt(offset + VALUE_OFFSET);
+    final byte[] blob = new byte[length];
+    final int oldPosition = buffer.position();
+    buffer.position(offset + VALUE_OFFSET + Integer.BYTES);
+    buffer.get(blob, 0, length);
+    buffer.position(oldPosition);
+    return blob;
+  }
+
+  /**
+   * Write a variably lengthed Long[] value to a {@link ByteBuffer} at the supplied offset. The first byte is set to
+   * {@link NullHandling#IS_NULL_BYTE} or {@link NullHandling#IS_NOT_NULL_BYTE} as appropriate, and if the Long[] value
+   * is not null, the size in bytes is written as an integer in the next 4 bytes. Elements of the array are each written
+   * out with {@link #writeNull} if null, or {@link #writeNullableLong} if not, taking either 1 or 9 bytes each. If the
+   * total byte size of serializing the array is larger than the max size parameter, this method will explode via a call
+   * to {@link #checkMaxBytes}.
+   *
+   * layout: | null (byte) | size (int) | {| null (byte) | long |, | null (byte) |, ... |null (byte) | long |} |
+   *
+   *
+   * This method does not permanently change the buffer position, limit, or mark.
+   *
+   * @return number of bytes written (1 if null, or 5 + size of Long[] if not)
+   */
+  public static int writeNullableLongArray(ByteBuffer buffer, int offset, @Nullable Long[] array, int maxSizeBytes)
+  {
+    // | null (byte) | array length (int) | array bytes |
+    if (array == null) {
+      return writeNull(buffer, offset);
+    } else {
+      int sizeBytes = 1 + Integer.BYTES;
+
+      buffer.put(offset, NullHandling.IS_NOT_NULL_BYTE);
+      buffer.putInt(offset + 1, array.length);
+      for (Long element : array) {
+        if (element != null) {
+          checkMaxBytes(
+              ColumnType.LONG_ARRAY,
+              sizeBytes + 1 + Long.BYTES,
+              maxSizeBytes
+          );
+          sizeBytes += writeNullableLong(buffer, offset + sizeBytes, element);
+        } else {
+          checkMaxBytes(
+              ColumnType.LONG_ARRAY,
+              sizeBytes + 1,
+              maxSizeBytes
+          );
+          sizeBytes += writeNull(buffer, offset + sizeBytes);
+        }
+      }
+      return sizeBytes;
+    }
+  }
+
+  /**
+   * Reads a nullable variably lengthed Long[] value from a {@link ByteBuffer} at the supplied offset. If the null byte
+   * is set to {@link NullHandling#IS_NULL_BYTE}, this method will return null, else it will read the size of the array
+   * from the next 4 bytes and then read that many elements with {@link #isNullableNull} and {@link #readNullableLong}.
+   *
+   * layout: | null (byte) | size (int) | {| null (byte) | long |, | null (byte) |, ... |null (byte) | long |} |
+   *
+   * This method does not change the buffer position, limit, or mark.
```

Comment:
Why? The byte buffer has a nice mechanism to keep track of the write mechanism. These functions recreate that by taking a write position, computing the amount of data written, returning that value, and asking the caller to compute the new write offset. These functions write variable amounts of data, so that there is only one right place to write the next value: after the current one.

Maybe explain why we need to recreate the byte buffer write pointer?

Reply:
Ah these methods are not really well suited for writing some sequence of values, despite that the array readers/writers use them that way. These methods are currently built with the buffer aggregators in mind, where these methods do not own the buffer that they operate on, instead they are used to read and write values at specific positions within a shared buffer that are associated with some key. These methods also happen at high volume, reading and writing values to different offsets in the buffer potentially every row, so the overhead of duping buffers with clean positions and limits seems non-optimal.

I did change stuff around a bit to lean more into temporarily setting and resetting the position for some of the variably sized things and using that to compute sizes.